MapboxCoreNavigation/RouteController.swift

```diff
@@ -193,7 +193,8 @@ open class RouteController: NSObject {
         
         locations.forEach { navigator.updateLocation(for: FixLocation($0)) }
         
-        let status = navigator.status(at: location.timestamp)
+        let timestampForStatus = Date().addingTimeInterval(RouteControllerDeadReckoningTimeInterval)
+        let status = navigator.status(at: timestampForStatus)
```

Comment:
I’m concerned that we’re taking a blanket approach to fast-forwarding location tracking based solely on the map’s built-in animation behavior. The motivation behind this change was to align the camera and puck’s animated position on screen with the user’s real-world location, accounting for the built-in 1&nbsp;second animation. However, other mechanisms depend on these location updates, namely guidance instructions, rerouting, arrival, tunnel-related style switching, and any other UI or business logic at the application level.

The ideal solution would be to move this lookahead logic to MapboxNavigation: `NavigationViewController.navigationService(_:didUpdate:with:rawLocation:)` would ignore the passed-in locations (which would not be influenced by any lookahead) in favor of calling `RouteController.projectedLocation(for:)`, passing in a lookahead that matches the animation duration. This would more appropriately keep UI-related concerns together:

https://github.com/mapbox/mapbox-navigation-ios/blob/ebab7d0881d33a975e265e8db20ca88d2eb7918a/MapboxNavigation/NavigationViewController.swift#L546 https://github.com/mapbox/mapbox-navigation-ios/blob/ebab7d0881d33a975e265e8db20ca88d2eb7918a/MapboxCoreNavigation/RouteController.swift#L340-L346

There are two downsides to this approach:

* If the application implements a custom UI instead of using NavigationViewController, it’s the application’s responsibility to call `RouteController.projectedLocation(for:)`, which may not be obvious. We can address this expectation in documentation and examples.
* More concerningly, `RouteController.projectedLocation(for:)` must call `Navigator.status(at:)` directly; calling `Navigator.status(at:)` out of band causes the navigator to skip instructions and potentially behave nondeterministically: #2113. I don’t think the workaround in https://github.com/mapbox/mapbox-navigation-ios/issues/2113#issuecomment-513833620 would help at all, because we would still need two different statuses at any given time. @SiarheiFedartsou, do you know if the underlying MapboxNavigationNative issue has been resolved?

Reply:
> The ideal solution would be to move this lookahead logic to MapboxNavigation: `NavigationViewController.navigationService(_:didUpdate:with:rawLocation:)` would ignore the passed-in locations (which would not be influenced by any lookahead) in favor of calling `RouteController.projectedLocation(for:)`, passing in a lookahead that matches the animation duration.

#2583 tracks removing `RouteController.projectedLocation(for:)`, because MapboxNavigationNative won’t support this (already buggy) functionality for much longer. #2587 tracks decoupling the actual and projected location so that the map view can animate the camera and puck to the right target position without affecting non-animated components.